File: api/MCP/tools/approvals.py
"""
Approval workflow-related tools mixin.
"""
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class GetAvailableApproversArgs(BaseModel):
    pass  # No arguments needed


class SubmitExpenseForApprovalArgs(BaseModel):
    expense_id: int = Field(description="ID of the expense to submit")
    notes: Optional[str] = Field(default=None, description="Optional submission notes")
    approver_id: Optional[int] = Field(default=None, description="Optional specific approver ID")


class GetPendingApprovalsArgs(BaseModel):
    limit: Optional[int] = Field(default=None, ge=1, le=100, description="Maximum number of results")
    offset: Optional[int] = Field(default=None, ge=0, description="Number of results to skip")


class GetPendingApprovalsSummaryArgs(BaseModel):
    pass  # No arguments needed


class ApproveExpenseArgs(BaseModel):
    approval_id: int = Field(description="ID of the approval to approve")
    notes: Optional[str] = Field(default=None, description="Optional approval notes")


class RejectExpenseArgs(BaseModel):
    approval_id: int = Field(description="ID of the approval to reject")
    rejection_reason: str = Field(description="Required rejection reason")
    notes: Optional[str] = Field(default=None, description="Optional additional notes")


class GetApprovalHistoryArgs(BaseModel):
    expense_id: int = Field(description="ID of the expense")


class GetApprovalMetricsArgs(BaseModel):
    approver_id: Optional[int] = Field(default=None, description="Filter by specific approver ID")


class GetApprovalDashboardStatsArgs(BaseModel):
    pass  # No arguments needed


class GetApprovedExpensesArgs(BaseModel):
    skip: int = Field(default=0, ge=0, description="Number of records to skip")
    limit: int = Field(default=50, ge=1, le=100, description="Maximum number of records")


class GetProcessedExpensesArgs(BaseModel):
    skip: int = Field(default=0, ge=0, description="Number of records to skip")
    limit: int = Field(default=50, ge=1, le=100, description="Maximum number of records")


class CreateApprovalDelegationArgs(BaseModel):
    delegate_id: int = Field(description="ID of the user to delegate to")
    start_date: str = Field(description="Delegation start date (YYYY-MM-DD)")
    end_date: str = Field(description="Delegation end date (YYYY-MM-DD)")
    is_active: bool = Field(default=True, description="Whether delegation is active")


class GetApprovalDelegationsArgs(BaseModel):
    include_inactive: bool = Field(default=False, description="Include inactive delegations")


class UpdateApprovalDelegationArgs(BaseModel):
    delegation_id: int = Field(description="ID of the delegation to update")
    start_date: Optional[str] = Field(default=None, description="New start date")
    end_date: Optional[str] = Field(default=None, description="New end date")
    is_active: Optional[bool] = Field(default=None, description="New active status")


class DeactivateApprovalDelegationArgs(BaseModel):
    delegation_id: int = Field(description="ID of the delegation to deactivate")


class ApprovalToolsMixin:
    # === Approval Management Tools (later, more complete versions) ===

    async def get_available_approvers(self) -> Dict[str, Any]:
        """Get list of available approvers"""
        try:
            response = await self.api_client._make_request("GET", "/approvals/approvers")
            return {
                "success": True,
                "data": response,
                "count": len(response),
                "message": f"Found {len(response)} available approvers"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to get available approvers: {e}"}

    async def submit_expense_for_approval(self, expense_id: int, notes: Optional[str] = None, approver_id: Optional[int] = None) -> Dict[str, Any]:
        """Submit an expense for approval"""
        try:
            payload = {"expense_id": expense_id}
            if notes:
                payload["notes"] = notes
            if approver_id:
                payload["approver_id"] = approver_id

            response = await self.api_client._make_request("POST", f"/approvals/expenses/{expense_id}/submit-approval", json=payload)
            return {
                "success": True,
                "data": response,
                "message": "Expense submitted for approval successfully"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to submit expense for approval: {e}"}

    async def get_pending_approvals(self, limit: Optional[int] = None, offset: Optional[int] = None) -> Dict[str, Any]:
        """Get pending approvals for current user"""
        try:
            params = {}
            if limit is not None:
                params["limit"] = limit
            if offset is not None:
                params["offset"] = offset

            response = await self.api_client._make_request("GET", "/approvals/pending", params=params)

            # Extract approvals from response
            approvals = self._extract_items_from_response(response, ["approvals", "items", "data"])

            return {
                "success": True,
                "data": approvals,
                "total": response.get("total", 0) if isinstance(response, dict) else len(approvals),
                "message": f"Found {response.get('total', 0) if isinstance(response, dict) else len(approvals)} pending approvals"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to get pending approvals: {e}"}

    async def get_pending_approvals_summary(self) -> Dict[str, Any]:
        """Get summary of pending approvals"""
        try:
            response = await self.api_client._make_request("GET", "/approvals/pending/summary")
            return {
                "success": True,
                "data": response,
                "message": "Pending approvals summary retrieved"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to get pending approvals summary: {e}"}

    async def approve_expense(self, approval_id: int, notes: Optional[str] = None) -> Dict[str, Any]:
        """Approve an expense"""
        try:
            payload = {"status": "approved"}
            if notes:
                payload["notes"] = notes

            response = await self.api_client._make_request("POST", f"/approvals/{approval_id}/approve", json=payload)
            return {
                "success": True,
                "data": response,
                "message": "Expense approved successfully"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to approve expense: {e}"}

    async def reject_expense(self, approval_id: int, rejection_reason: str, notes: Optional[str] = None) -> Dict[str, Any]:
        """Reject an expense"""
        try:
            payload = {
                "status": "rejected",
                "rejection_reason": rejection_reason
            }
            if notes:
                payload["notes"] = notes

            response = await self.api_client._make_request("POST", f"/approvals/{approval_id}/reject", json=payload)
            return {
                "success": True,
                "data": response,
                "message": "Expense rejected successfully"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to reject expense: {e}"}

    async def get_approval_history(self, expense_id: int) -> Dict[str, Any]:
        """Get approval history for an expense"""
        try:
            response = await self.api_client._make_request("GET", f"/approvals/history/{expense_id}")
            return {
                "success": True,
                "data": response,
                "message": "Approval history retrieved successfully"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to get approval history: {e}"}

    async def get_approval_metrics(self, approver_id: Optional[int] = None) -> Dict[str, Any]:
        """Get approval workflow metrics"""
        try:
            params = {}
            if approver_id:
                params["approver_id"] = approver_id

            response = await self.api_client._make_request("GET", "/approvals/metrics", params=params)
            return {
                "success": True,
                "data": response,
                "message": "Approval metrics retrieved successfully"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to get approval metrics: {e}"}

    async def get_approval_dashboard_stats(self) -> Dict[str, Any]:
        """Get approval dashboard statistics"""
        try:
            response = await self.api_client._make_request("GET", "/approvals/dashboard-stats")
            return {
                "success": True,
                "data": response,
                "message": "Dashboard statistics retrieved successfully"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to get dashboard stats: {e}"}

    async def get_approved_expenses(self, skip: int = 0, limit: int = 50) -> Dict[str, Any]:
        """Get expenses approved by current user"""
        try:
            params = {"skip": skip, "limit": limit}
            response = await self.api_client._make_request("GET", "/approvals/approved-expenses", params=params)

            # Extract expenses from response
            expenses = self._extract_items_from_response(response, ["expenses", "items", "data"])

            return {
                "success": True,
                "data": expenses,
                "total": response.get("total", 0) if isinstance(response, dict) else len(expenses),
                "message": f"Found {response.get('total', 0) if isinstance(response, dict) else len(expenses)} approved expenses"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to get approved expenses: {e}"}

    async def get_processed_expenses(self, skip: int = 0, limit: int = 50) -> Dict[str, Any]:
        """Get expenses processed (approved/rejected) by current user"""
        try:
            params = {"skip": skip, "limit": limit}
            response = await self.api_client._make_request("GET", "/approvals/processed-expenses", params=params)

            # Extract expenses from response
            expenses = self._extract_items_from_response(response, ["expenses", "items", "data"])

            return {
                "success": True,
                "data": expenses,
                "total": response.get("total", 0) if isinstance(response, dict) else len(expenses),
                "message": f"Found {response.get('total', 0) if isinstance(response, dict) else len(expenses)} processed expenses"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to get processed expenses: {e}"}

    async def create_approval_delegation(self, delegate_id: int, start_date: str, end_date: str, is_active: bool = True) -> Dict[str, Any]:
        """Create an approval delegation"""
        try:
            payload = {
                "delegate_id": delegate_id,
                "start_date": start_date,
                "end_date": end_date,
                "is_active": is_active
            }
            response = await self.api_client._make_request("POST", "/approvals/delegate", json=payload)
            return {
                "success": True,
                "data": response,
                "message": "Approval delegation created successfully"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to create approval delegation: {e}"}

    async def get_approval_delegations(self, include_inactive: bool = False) -> Dict[str, Any]:
        """Get approval delegations for current user"""
        try:
            params = {"include_inactive": include_inactive}
            response = await self.api_client._make_request("GET", "/approvals/delegates", params=params)
            return {
                "success": True,
                "data": response,
                "count": len(response),
                "message": f"Found {len(response)} delegations"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to get approval delegations: {e}"}

    async def update_approval_delegation(self, delegation_id: int, start_date: Optional[str] = None, end_date: Optional[str] = None, is_active: Optional[bool] = None) -> Dict[str, Any]:
        """Update an approval delegation"""
        try:
            payload = {}
            if start_date:
                payload["start_date"] = start_date
            if end_date:
                payload["end_date"] = end_date
            if is_active is not None:
                payload["is_active"] = is_active

            response = await self.api_client._make_request("PUT", f"/approvals/delegates/{delegation_id}", json=payload)
            return {
                "success": True,
                "data": response,
                "message": "Approval delegation updated successfully"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to update approval delegation: {e}"}

    async def deactivate_approval_delegation(self, delegation_id: int) -> Dict[str, Any]:
        """Deactivate an approval delegation"""
        try:
            response = await self.api_client._make_request("DELETE", f"/approvals/delegates/{delegation_id}")
            return {
                "success": True,
                "data": response,
                "message": "Approval delegation deactivated successfully"
            }
        except Exception as e:
            return {"success": False, "error": f"Failed to deactivate approval delegation: {e}"}

File: ui/src/lib/api/approvals.ts
import { apiRequest } from './_base';
import type { ExpenseApproval, ApprovalHistoryEntry, ApprovalDashboardStats, ApprovalDelegate, ApprovalDelegateCreate, ApprovalDelegateUpdate } from '@/types';
import type { Expense } from './expenses';

// Approval API methods
export const approvalApi = {
  // Get pending approvals for current user
  getPendingApprovals: (filters?: {
    limit?: number;
    offset?: number;
    category?: string;
    min_amount?: number;
    max_amount?: number;
    sort_by?: string;
    sort_order?: 'asc' | 'desc';
  }) => {
    const params = new URLSearchParams();
    if (filters?.limit !== undefined) params.append('limit', filters.limit.toString());
    if (filters?.offset !== undefined) params.append('offset', filters.offset.toString());
    if (filters?.category) params.append('category', filters.category);
    if (filters?.min_amount) params.append('min_amount', filters.min_amount.toString());
    if (filters?.max_amount) params.append('max_amount', filters.max_amount.toString());
    if (filters?.sort_by) params.append('sort_by', filters.sort_by);
    if (filters?.sort_order) params.append('sort_order', filters.sort_order);

    const queryString = params.toString();
    return apiRequest<{ approvals: ExpenseApproval[]; total: number; }>(`/approvals/pending${queryString ? `?${queryString}` : ''}`);
  },

  // Get approval dashboard statistics
  getDashboardStats: () => apiRequest<ApprovalDashboardStats>("/approvals/dashboard-stats"),

  // Get expenses approved by current user
  getApprovedExpenses: (filters?: {
    skip?: number;
    limit?: number;
  }) => {
    const params = new URLSearchParams();
    if (filters?.skip !== undefined) params.append('skip', filters.skip.toString());
    if (filters?.limit !== undefined) params.append('limit', filters.limit.toString());

    const queryString = params.toString();
    return apiRequest<{ expenses: Expense[]; total: number; }>(`/approvals/approved-expenses${queryString ? `?${queryString}` : ''}`);
  },

  // Get expenses processed (approved/rejected) by current user
  getProcessedExpenses: (filters?: {
    skip?: number;
    limit?: number;
  }) => {
    const params = new URLSearchParams();
    if (filters?.skip !== undefined) params.append('skip', filters.skip.toString());
    if (filters?.limit !== undefined) params.append('limit', filters.limit.toString());

    const queryString = params.toString();
    return apiRequest<{ expenses: Expense[]; total: number; }>(`/approvals/processed-expenses${queryString ? `?${queryString}` : ''}`);
  },

  // Get invoices processed (approved/rejected) by current user
  getProcessedInvoices: (filters?: {
    skip?: number;
    limit?: number;
  }) => {
    const params = new URLSearchParams();
    if (filters?.skip !== undefined) params.append('skip', filters.skip.toString());
    if (filters?.limit !== undefined) params.append('limit', filters.limit.toString());

    const queryString = params.toString();
    return apiRequest<{ invoices: any[]; total: number; }>(`/approvals/invoices/processed${queryString ? `?${queryString}` : ''}`);
  },

  // Approve an expense
  approveExpense: (approvalId: number, notes?: string) =>
    apiRequest<{ success: boolean; message: string; }>(`/approvals/${approvalId}/approve`, {
      method: 'POST',
      body: JSON.stringify({ status: 'approved', notes }),
    }),

  // Reject an expense
  rejectExpense: (approvalId: number, reason: string, notes?: string) =>
    apiRequest<{ success: boolean; message: string; }>(`/approvals/${approvalId}/reject`, {
      method: 'POST',
      body: JSON.stringify({ status: 'rejected', rejection_reason: reason, notes }),
    }),

  // Unsubmit an expense approval request
  unsubmitExpenseApproval: (expenseId: number) =>
    apiRequest<{ success: boolean; message: string; }>(`/approvals/expenses/${expenseId}/unsubmit`, {
      method: 'POST',
    }),

  // Get approval history for an expense
  getApprovalHistory: (expenseId: number) =>
    apiRequest<{ history: ApprovalHistoryEntry[]; }>(`/approvals/history/${expenseId}`),

  // Submit expense for approval
  submitForApproval: (expenseId: number, approverId: number, notes?: string) =>
    apiRequest<ExpenseApproval[]>(`/approvals/expenses/${expenseId}/submit-approval`, {
      method: 'POST',
      body: JSON.stringify({ expense_id: expenseId, notes, approver_id: approverId }),
    }),

  // Get list of available approvers
  getApprovers: () => apiRequest<Array<{ id: number; name: string; email: string }>>(`/approvals/approvers`),

  // Invoice Approval Methods
  submitInvoiceForApproval: (invoiceId: number, data: { approver_id: number; notes?: string }) =>
    apiRequest<Array<{ id: number; status: string; approval_level: number }>>(`/approvals/invoices/${invoiceId}/submit-approval`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  getPendingInvoiceApprovals: (filters?: { limit?: number; offset?: number }) => {
    const params = new URLSearchParams();
    if (filters?.limit !== undefined) params.append('limit', filters.limit.toString());
    if (filters?.offset !== undefined) params.append('offset', filters.offset.toString());

    const queryString = params.toString();
    return apiRequest<{ approvals: any[]; total: number; }>(`/approvals/invoices/pending${queryString ? `?${queryString}` : ''}`);
  },

  approveInvoice: (approvalId: number, notes?: string) =>
    apiRequest<{ id: number; status: string; invoice_id: number }>(`/approvals/invoices/${approvalId}/approve`, {
      method: 'POST',
      body: JSON.stringify({ notes }),
    }),

  getInvoiceApprovalHistory: (invoiceId: number) =>
    apiRequest<{ invoice_id: number; current_status: string; approval_history: ApprovalHistoryEntry[]; }>(`/approvals/invoices/history/${invoiceId}`),

  rejectInvoice: (approvalId: number, reason: string, notes?: string) =>
    apiRequest<{ id: number; status: string; invoice_id: number }>(`/approvals/invoices/${approvalId}/reject`, {
      method: 'POST',
      body: JSON.stringify({ rejection_reason: reason, notes }),
    }),

  // Unsubmit an invoice approval request
  unsubmitInvoiceApproval: (invoiceId: number) =>
    apiRequest<{ success: boolean; message: string; }>(`/approvals/invoices/${invoiceId}/unsubmit`, {
      method: 'POST',
    }),



  // Approval Delegation Management
  getDelegations: (filters?: {
    approver_id?: number;
    delegate_id?: number;
    is_active?: boolean;
    limit?: number;
    offset?: number;
  }) => {
    const params = new URLSearchParams();
    if (filters?.approver_id) params.append('approver_id', filters.approver_id.toString());
    if (filters?.delegate_id) params.append('delegate_id', filters.delegate_id.toString());
    if (filters?.is_active !== undefined) params.append('is_active', filters.is_active.toString());
    if (filters?.limit !== undefined) params.append('limit', filters.limit.toString());
    if (filters?.offset !== undefined) params.append('offset', filters.offset.toString());

    const queryString = params.toString();
    return apiRequest<ApprovalDelegate[]>(`/approvals/delegates${queryString ? `?${queryString}` : ''}`);
  },

  createDelegation: (delegationData: ApprovalDelegateCreate) =>
    apiRequest<ApprovalDelegate>('/approvals/delegate', {
      method: 'POST',
      body: JSON.stringify(delegationData),
    }),

  updateDelegation: (delegationId: number, delegationData: ApprovalDelegateUpdate) =>
    apiRequest<ApprovalDelegate>(`/approvals/delegates/${delegationId}`, {
      method: 'PUT',
      body: JSON.stringify(delegationData),
    }),

  deleteDelegation: (delegationId: number) =>
    apiRequest<{ message: string }>(`/approvals/delegates/${delegationId}`, {
      method: 'DELETE',
    }),
};
